pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -758,7 +758,23 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
     }
   }
 
-  protected abstract void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds);
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
+    doRemoveSegment(segment, doGetRecordInfoIteratorForRemoveSegment(segment, validDocIds));
+  }
+
+  protected Iterator<RecordInfo> doGetRecordInfoIteratorForRemoveSegment(IndexSegment segment,
+      MutableRoaringBitmap validDocIds) {
+    try (UpsertUtils.RecordInfoReader recordInfoReader = new UpsertUtils.RecordInfoReader(segment, _primaryKeyColumns,
+        _comparisonColumns, _deleteRecordColumn)) {
+      return UpsertUtils.getRecordInfoIterator(recordInfoReader, validDocIds);
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Caught exception while removing segment: %s, table: %s", segment.getSegmentName(),
+              _tableNameWithType), e);
+    }
+  }
+
+  protected abstract void doRemoveSegment(IndexSegment segment, Iterator<RecordInfo> recordInfoIterator);
```

Comment:
Suggest not adding this method because `RecordInfo` is used for adding record, instead of removing record. It contains unnecessary info for record removal which can cause inefficiency.

Reply:
Thanks as discussed offline, we can convert this to a primary-key iterator.